### PR TITLE
Refactor: Use Constants

### PR DESCRIPTION
**Pull request by strehle**
RFC 7517 constants from jose library

### CODE REVIEW

File: model/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKey.java [hunk 1]
```diff
@@ -126,32 +127,32 @@ public int hashCode() {
...
+        List<String> result = (List<String>) getKeyProperties().get(JWKParameterNames.KEY_OPS);
         if (result==null) {
             result = Collections.emptyList();
         }
-        return result.stream().map(o -> KeyOperation.valueOf(o)).collect(Collectors.toSet());
+        return result.stream().map(KeyOperation::valueOf).collect(Collectors.toSet());
     }
 
-    public static String pemEncodePublicKey(PublicKey publicKey) {
+    public static String pemEncodePublicKey(Key publicKey) {
```
Comment (bruce-ricard) [inline model/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKey.java]: This one feels worse.

Now the method is called `pemEncode` **publicKey**, but the argument is just a `Key`. I don't like this name. What kind of key is it? The key to my bike lock?

File: model/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKey.java [hunk 2]
```diff
@@ -126,32 +127,32 @@ public int hashCode() {
...
     }
 
     public Set<KeyOperation> getKeyOps() {
-        List<String> result = (List<String>) getKeyProperties().get("key_ops");
+        List<String> result = (List<String>) getKeyProperties().get(JWKParameterNames.KEY_OPS);
         if (result==null) {
             result = Collections.emptyList();
         }
-        return result.stream().map(o -> KeyOperation.valueOf(o)).collect(Collectors.toSet());
+        return result.stream().map(KeyOperation::valueOf).collect(Collectors.toSet());
```
Comment (bruce-ricard) [inline model/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKey.java:151]: Nice!

File: model/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKey.java [hunk 3]
```diff
@@ -161,12 +162,12 @@ public static String pemEncodePublicKey(PublicKey publicKey) {
 
     public static PublicKey getRsaPublicKey(JsonWebKey key) {
         final Base64 decoder = new Base64(true);
-        String e = (String) key.getKeyProperties().get("e");
-        String n = (String) key.getKeyProperties().get("n");
+        String e = (String) key.getKeyProperties().get(JWKParameterNames.RSA_EXPONENT);
+        String n = (String) key.getKeyProperties().get(JWKParameterNames.RSA_MODULUS);
```
Comment (bruce-ricard) [inline model/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKey.java:165]: Awesome.

File: server/src/test/java/org/cloudfoundry/identity/uaa/provider/oauth/ExternalOAuthAuthenticationManagerTest.java [hunk 4]
```diff
@@ -158,8 +157,8 @@ public void getExternalAuthenticationDetails_whenProviderHasSigningKey_throwsWhe
         expectedException.expectMessage("Could not verify token signature.");
 
         Map<String, Object> header = map(
-                entry(ALG, "HS256"),
-                entry(KID, "oidc-provider-key")
+                entry(HeaderParameterNames.ALGORITHM, "HS256"),
+                entry(HeaderParameterNames.KEY_ID, "oidc-provider-key")
```
Comment (bruce-ricard) [inline server/src/test/java/org/cloudfoundry/identity/uaa/provider/oauth/ExternalOAuthAuthenticationManagerTest.java]: It feels like the 2 values "HS256" and "oidc-provider-key" could also be made constants.

**Review by bruce-ricard — APPROVED** [review]
Great commit. Thank you Markus. I've been waiting for this for years :).

I just made a couple suggestions inline if you are interested.

**Comment by bruce-ricard** [comment]
A suggestion too: instead of having a commit message saying "more refactorings", you could give just a tiny bit of extra information. For example "refactor: use constants" would be my choice for a commit like [this one](https://github.com/cloudfoundry/uaa/pull/2181/commits/461c122f6ca92da1c1762fe00576598fd2863c36)

**Comment by Tallicia** [comment]
This looks good to me.

File: model/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKey.java (hunk 1 above)
Comment (strehle) [reply model/src/main/java/org/cloudfoundry/identity/uaa/oauth/jwk/JsonWebKey.java]: Sonar suggested this to me

File: server/src/test/java/org/cloudfoundry/identity/uaa/provider/oauth/ExternalOAuthAuthenticationManagerTest.java (hunk 4 above)
Comment (strehle) [reply server/src/test/java/org/cloudfoundry/identity/uaa/provider/oauth/ExternalOAuthAuthenticationManagerTest.java]: done

**Pull request merged**